drivers: flash: nrf_qspi_nor: port downstream fixes for mcumgr support

pabigot opened the pull request:
Two changes initially made downstream and now ported upstream to fix failures in mcumgr capabilities due to the Nordic QSPI peripheral requiring aligned transfers of 32-bit words to/from SRAM:

* If an aligned write is requested but the transfer size is not a multiple of 4 then read the current word content, replace the prefix that is to change, and write back the whole word.
* If a write is requested from a pointer to NVMC, and a Kconfig option enables it, perform the write by copying blocks from NVMC to a stack-allocated buffer.

Inspired by commits in nrfconnect/sdk-zephyr#305 though significantly reworked for upstream.

pabigot committed:
drivers: flash: nrf_qspi_nor: support write of sub-word lengths

mcumgr and possibly mcuboot write single byte values to update the
state of objects.  Rather than fail to do the write of values too
short for this peripheral detect the situation and write from a stack
buffer that meets the length criteria.

Signed-off-by: Sigvart Hovland <sigvart.m@gmail.com>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

sigvartmh reviewed (dismissed):
To me this looks Good :)

rerickson1 commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -574,11 +575,33 @@ static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
+	}
+
+	if (res == NRFX_SUCCESS) {
+		res = nrfx_qspi_write(src, size, addr);
```
Shouldnt there be `qspi_wait_for_completion(dev, res);` after this write?

pabigot committed:
drivers: flash: nrf_qspi_nor: optionally support write from NVMC

The Nordic QSPI peripheral uses DMA transfers so data to write must be
located in SRAM.  Add a Kconfig that enables copying data from NVMC to
a stack SRAM buffer so it can be written to flash.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Updated with a fix to support writing NVMC-stored boot records, which must be enabled in both mcuboot and the Zephyr application containing mcumgr.

For mcuboot see: https://github.com/pabigot/mcuboot/commits/issue/25294
For zephyr see: https://github.com/pabigot/zephyr/commits/issue/25294

sigvartmh commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -540,15 +540,71 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
+BUILD_ASSERT((CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE % 4) == 0,
+	     "NOR stack buffer must be multiple of 4 bytes");
+
+#define NVMC_WRITE_OK ((CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE - 0) > 0)
```
I'm a little bit confused by this line. Care to elaborate why you need the `- 0` ?

pabigot replied:
In this case we probably don't.  One would use `(SYM - 0)` in a situation where `SYM` may be undefined, because to the preprocessor it would evaluate to `( - 0)` making the condition valid and false.

Here I don't think the module can see a case where `CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE` is not defined (and if it did the `BUILD_ASSERT` above would fail.  Gone now.

sigvartmh reviewed (approved):
Small comment otherwise it looks okei to me.

carlescufi commented:
@rerickson1 can you take another look?

rerickson1 reviewed (approved):
Looks good.
